docs/ts/primitives/streaming-apis.mdx

```diff
@@ -8,118 +8,269 @@ lang: ts
 
 If you don't need any data from the handshake, you can ignore that type, and only specify the incoming and outgoing message types.
 
-For example, if you want to have a stream of messages from server to client for a specific chat room, you could setup an `api.streamOut`:
+<GitHubLink
+    href="https://github.com/encoredev/examples/tree/main/ts/streaming"
+    desc="Example showcases the all different streaming APIs: api.streamIn, api.streamOut, and api.streamInOut"
+/>
+
+### StreamIn
+
+Use `api.streamIn` when you want to have a stream from client to server, for example if you are uploading something from the client to the server:
 
 ```typescript
+import { api } from "encore.dev/api";
+
+// Used to pass initial data, optional.
 interface Handshake {
-  roomId: string;
+  user: string;
 }
 
+// What the clients sends over the stream.
 interface Message {
-  text: string;
-  sender: string;
+  data: string;
+  done: boolean;
 }
 
-// messageStream streams out all chat messages for a certain id
-export const messageStream = api.streamOut<Handshake, Message>(
-  { path: "/chat/:roomId", expose: true },
-  async (handshake, stream) => {
-    // Use `handshake.roomId` to fetch messages
-    for await (const msg of messages) {
-      await stream.send(msg);
+// Returned when the stream is done, optional.
+interface Response {
+  success: boolean;
+}
+
+export const uploadStream = api.streamIn<
+  Handshake,
+  Message,
+  Response
+>({ path: "/upload", expose: true }, async (handshake, stream) => {
+  const chunks: string[] = [];
+  try {
+    // The stream object is an AsyncIterator that yields incoming messages.
+    for await (const data of stream) {
+      chunks.push(data.data);
+      // Stop the stream if the client sends a "done" message
+      if (data.done) break;
     }
+  } catch (err) {
+    console.error(`Upload error by ${handshake.user}:`, err);
+    return { success: false };
   }
-);
+  console.log(`Upload complete by ${handshake.user}:`, chunks);
+  return { success: true };
+});
+
+```
+
+For `api.streamIn` you need to specify the incoming message type, the handshake type is optional. You can also specify a optional outgoing type if your API handler responds with some data when it is done with the incoming stream.
+
+```ts
+api.streamIn<Handshake, Incoming, Outgoing>(
+  {...}, async (handshake, stream): Promise<Outgoing> => {...})
 ```
 
-Or if you want to have a stream from client to server, you can set it up with `api.streamIn`:
+```ts
+api.streamIn<Handshake, Incoming>(
+  {...}, async (handshake, stream) => {...})
+```
+
+```ts
+api.streamIn<Incoming, Outgoing>(
+  {...}, async (stream): Promise<Outgoing> => {...})
```

Comment:
I guess we don't need the promise here either

Reply:
I thought it was clearer when it was explicit, no?

Reply:
Seems like extra typing to me :) in most TypeScript code I've seen people prefer to use type inference when possible. But I don't feel strongly about it, keep it if you prefer